source/ImageSet.cpp

```diff
@@ -45,6 +47,44 @@ namespace {
 		// are part of the sprite name, not a frame index.
 		return (IsBlend(path[pos]) ? pos : end);
 	}
+
+	// Log an error if frames are missing in one of the paths vectors.
+	void LogIfMissingFrames(const vector<string>& paths, size_t expectedCount, string prefix, bool is2x)
+	{
+		size_t firstGapStart = paths.size();
+		size_t firstGapEnd = expectedCount - 1;
+		const size_t totalFrames = min(paths.size(), expectedCount);
+		size_t totalMissing = expectedCount - totalFrames;
+		for(size_t i = 0; i < totalFrames; ++i)
+		{
+			if(paths[i].empty())
+			{
+				totalMissing++;
+				firstGapStart = min(firstGapStart, i);
+			}
+			else if(firstGapStart != paths.size() && firstGapEnd == expectedCount - 1)
+			{
+				firstGapEnd = i - 1;
+			}
+		}
+		if(!totalMissing)
+			return;
+
+		string message = prefix + "missing " + (is2x ? "@2x " : "");
+		if(firstGapStart == firstGapEnd)
+		{
+			message += "frame " + to_string(firstGapStart);
+		}
+		else
+		{
+			message += "frames " + to_string(firstGapStart) + "-" + to_string(firstGapEnd);
+		}
+		size_t extraMissing = totalMissing - (firstGapEnd - firstGapStart + 1);
+		if(extraMissing)
+			message += " and " + to_string(extraMissing) + " more";
```

Comment:
I think it's fine to not print the exact missing range. The content creator has their filesystem view and can use that to find gaps once they know one exists.

So e.g. "missing frame X (<totalmissing>)" is fine by me.

Reply:
That simplifies the code significnatly indeed